lib/web/conn_upgrade.go

```diff
@@ -88,6 +93,41 @@ func (h *Handler) upgradeALPN(ctx context.Context, conn net.Conn) error {
 	return h.cfg.ALPNHandler(ctx, waitConn)
 }
 
+func (h *Handler) upgradeALPNWithPing(ctx context.Context, conn net.Conn) error {
+	if h.cfg.ALPNHandler == nil {
+		return trace.BadParameter("missing ALPNHandler")
+	}
+
+	pingConn := pingconn.New(conn)
+
+	// Cancel ping background goroutine when connection is closed.
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	go h.startPing(ctx, pingConn)
+	time.Sleep(time.Millisecond * 100)
+
+	return h.upgradeALPN(ctx, pingConn)
+}
+
+func (h *Handler) startPing(ctx context.Context, pingConn *pingconn.PingConn) {
+	ticker := time.NewTicker(defaults.ProxyPingInterval)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-ticker.C:
+			err := pingConn.WritePing()
+			if err != nil {
+				if !utils.IsOKNetworkError(err) {
+					h.log.WithError(err).Warn("Failed to write ping message")
+				}
+				return
+			}
+		}
+	}
+}
```

Comment:
Can we really on the ALPN Proxy ping call: https://github.com/gravitational/teleport/blob/922fc8bfb4046f9617bed2ed007e57a53a7c40ca/lib/srv/alpnproxy/proxy.go#L398 and don't introduce additional places where ping protocol is handled   ? 
For instance we can extend the   condition for `handlerConn = p.handlePingConnection(ctx, tlsConn)` and check the connection type